core/src/Cuda/Kokkos_Cuda_Team.hpp

```diff
@@ -265,39 +267,42 @@ class CudaTeamMember {
     vector_reduce(reducer, reducer.reference());
   }
 
-  template <typename ReducerType>
-  KOKKOS_INLINE_FUNCTION static std::enable_if_t<is_reducer<ReducerType>::value>
-  vector_reduce(ReducerType const& reducer,
-                typename ReducerType::value_type& value) {
+  template <typename ReducerType, typename ValueType>
+  KOKKOS_INLINE_FUNCTION static void vector_reduce(ReducerType const& reducer,
+                                                   ValueType& value) {
     (void)reducer;
     (void)value;
-    KOKKOS_IF_ON_DEVICE(
-        (if (blockDim.x == 1) return;
-
-         // Intra vector lane shuffle reduction:
-         typename ReducerType::value_type tmp(value);
-         typename ReducerType::value_type tmp2 = tmp;
-
-         unsigned mask =
-             blockDim.x == 32
-                 ? 0xffffffff
-                 : ((1 << blockDim.x) - 1)
-                       << ((threadIdx.y % (32 / blockDim.x)) * blockDim.x);
-
-         for (int i = blockDim.x; (i >>= 1);) {
-           Impl::in_place_shfl_down(tmp2, tmp, i, blockDim.x, mask);
-           if ((int)threadIdx.x < i) {
-             reducer.join(tmp, tmp2);
-           }
-         }
-
-         // Broadcast from root lane to all other lanes.
-         // Cannot use "butterfly" algorithm to avoid the broadcast
-         // because floating point summation is not associative
-         // and thus different threads could have different results.
-
-         Impl::in_place_shfl(tmp2, tmp, 0, blockDim.x, mask);
-         value = tmp2; reducer.reference() = tmp2;))
+    KOKKOS_IF_ON_DEVICE((
+        if (blockDim.x == 1) return;
+
+        // Intra vector lane shuffle reduction:
+        typename ReducerType::value_type tmp(value);
+        typename ReducerType::value_type tmp2 = tmp;
+
+        unsigned mask =
+            blockDim.x == 32
+                ? 0xffffffff
+                : ((1 << blockDim.x) - 1)
+                      << ((threadIdx.y % (32 / blockDim.x)) * blockDim.x);
+
+        for (int i = blockDim.x; (i >>= 1);) {
+          Impl::in_place_shfl_down(tmp2, tmp, i, blockDim.x, mask);
+          if ((int)threadIdx.x < i) {
+            reducer.join(tmp, tmp2);
+          }
+        }
+
+        // Broadcast from root lane to all other lanes.
+        // Cannot use "butterfly" algorithm to avoid the broadcast
+        // because floating point summation is not associative
+        // and thus different threads could have different results.
+
+        Impl::in_place_shfl(tmp2, tmp, 0, blockDim.x, mask);
+        value = tmp2;
+
+        if constexpr (is_reducer_v<ReducerType>) {
+          reducer.reference() = tmp2;
+        }))
```

Comment:
Why aren't we using `final` here if defined?

Reply:
Fixed to call `final` through the `FunctorAnalysis::Reducer` interface.
It seems that `final` wasn't being called for any of nested team `parallel_reduce`.